drivers/gles3/rasterizer_scene_gles3.cpp

```diff
@@ -5152,8 +5152,29 @@ void RasterizerSceneGLES3::initialize() {
 	{
 		//spot and omni ubos
 
-		int max_ubo_size;
-		glGetIntegerv(GL_MAX_UNIFORM_BLOCK_SIZE, &max_ubo_size);
+		// SPECIAL CASE for GL_MAX_UNIFORM_BLOCK_SIZE.
+		// Under ANGLE, in some situations this will return INT32_MAX + 1.
+		// This seems to be because ANGLE supports system memory backing,
+		// but the true hardware GPU max may be lower, but we have no way
+		// of querying this hardware supported value.
+		// We also ideally want to take advantage of GPUs that *do* support large uniform
+		// blocks in hardware (although this is probably not taken advantage of in Godot 3.x currently).
+
+		// This logic is thus a compromise.
+		int max_ubo_size = RasterizerStorageGLES3::safe_gl_get_integer(GL_MAX_UNIFORM_BLOCK_SIZE);
+
+		// If UBO size if very large, we are likely under ANGLE and it is reporting system memory.
+		if (max_ubo_size >= INT32_MAX) {
```

Comment:
This is probably unnecessary. Just using the int64 is good enough and matches the spec. This is pretty complex, untested by defintion and an overreaction to the error that I precisely identified and rectified.

Reply:
Actually, this is incorrect. INT32_MAX is what my GPU reports on desktop, we should consider that valid.

Reply:
Removing this check will make the PR valid, but it's complexity obfuscates mistakes like this. IMO this PR is too aggressive. Godot 3 is a legacy project, it needs to be treated more conservatively.

I'd like to add that I vastly respect the work you've put in Lawn Jelly, and respect how willing you are to make ambitious changes.

Reply:
See https://github.com/godotengine/godot/pull/83031#issuecomment-1752690501 for reasoning.

It is admittedly controversial either way, there is no ideal, so I'm really waiting to see what @clayjohn thinks, as the discussion is relevant for 4.x.

I did waver between the two versions myself :thinking: , initially just using the large value as is, but as there is significant danger of blowing the GPU limit (in theory), I'm currently of the opinion that the reduction to the minimum in this case is the lesser of the evils. But either should be fine.

Reply:
Sounds good, will fix that up in the morning! :slightly_smiling_face: